text/1005-edgedb-project.rst

```diff
@@ -0,0 +1,418 @@
+
+The project -> instance mapping is expressed as directories under
+``~/.edgedb/projects``::
+
+    ~/.edgedb/projects/
+        <dir-basename>-<dir-abspath-hash>/
+            project-path
+            instance-name
+
+Here ``<dir-basename>`` is the trailing component of the path to the project
+directory, and ``<dir-abspath-hash>`` is defined as
+``lower(strip(base32(abspath(project_dir)), '==='))```.  Hashing is necessary
+to avoid bumping into the maximum directory entry name length.
+The ``project-path`` file contains the full absolute path to the project
+directory and is necessary for reverse lookup (find project by instance name),
+and the ``instance-name`` file contains the name of the associated instance.
+
+Language bindings and clients may cache the resolved instance name in memory
+to avoid performing a lookup on every ``connect()`` call.  The cache should
+be sensitive to the current working directory, although if the directory
+changed to a subdirectory the cache should still be valid as nested projects
+are not allowed.
+
+Running ``project init`` from ``curl | sh``
+-------------------------------------------
+
+The CLI bootstrap script can detect if it's being ran from within a project
+directory and ask if the user wants to run ``edgedb project init``.
+
+Effect on ``edgedb server`` commands
+------------------------------------
+
+The ``edgedb server destroy`` command should refuse to destroy an instance that
+is associated with a project by default, and should recommend to use
+``edgedb project deinit`` (or ``edgedb server destroy --force``).
+
+The ``edgedb server upgrade`` command should refuse to continue if the target
+server version does not match the ``edgedb.server-version`` range in
+``edgedb.toml``.
+
+
+edgedb project init
+===================
+
+Initialize a new project or re-initialize an existing project.
+
+Synopsis
+--------
+
+``edgedb project init [options]``
+
+Options
+-------
+
+``--project-dir=<dir>``
+  Specifies a project root directory explicitly.  If not specified, the project
+  directory is detected as described in the "Project detection" section above,
+  and if no project directory is detected, a current working directory is used.
+
+``--server-version=<semver-range>``
+  Specifies the desired EdgeDB server version as a SemVer range.  Only
+  applicable for new projects.  Defaults to ``'*'``, which means latest stable
+  version for new projects.  Accepts ``'nightly'`` as a special value denoting
+  the latest nightly version.
+
+``--server-instance=<instance>``
+  Specifies the EdgeDB server instance to be associated with the project.
+  If the specified instance does not exist, it will be created.  If the
+  specified instance already exists, it must not be associated with another
+  project.  ``edgedb server deinit`` may be used to disassociate an instance
+  prior to linking it with another project.
+
+``--server-instance-type=<instance-type>``
```

Comment:
Why `instance-type` and not `install-method` like we use elsewhere? Should we rename `server install --type` ?

Reply:
This is a bit different.  `install-method` refers to the local installation method (native vs docker), whereas instance type specifies `local` vs `cloud`, although in this case `local` branches to include the installation methods.  We can make a separate `--install-method` option instead and keep `--instance-type` as `{'local`, 'cloud'}`.

Reply:
I've split the option into two.